### DISCUSSION TITLE
Object storage: policy to restrict file size?

### DISCUSSION THREAD

**Discussion by Vinzius** [supabase/supabase#1123] [Questions]
Hey,

The new storage feature is very promising :-) congrats!

I was hoping to create a policy in order to restrict to file size. So avoid anyone uploading a big file in the S3 😅

I did try several things, the latest being:
```
CREATE POLICY "Any file: size should be less than 10MB." 
	ON storage.objects 
	FOR INSERT 
	WITH CHECK (
		(metadata->>'size')::integer < 10000000
	);
```
(I removed all the other policies to try it)

But I get the following error:
`{"statusCode":"42501","error":"","message":"new row violates row-level security policy for table \"objects\""}`

Am I doing something wrong here? I'm no Postgresql expert, so I may have missed something ^^

I did print the schemas of storage.objects just to be sure I didn't miss something.
<img width="1161" alt="Capture d’écran 2021-04-12 à 09 53 38" src="https://user-images.githubusercontent.com/337930/114362796-0825ae00-9b78-11eb-9237-b1a27f1f3d44.png">

And I tried a SQL query to fetch objects regarding their size:
`SELECT (metadata ->> 'size') FROM storage.objects WHERE (metadata ->> 'size')::integer < 1000000 LIMIT 10;`

So I guess it's not related to the table itself (ie wrong column format).

Thanks!

**Comment by inian** [comment]
Hi @Vinzius!

Are you trying to restrict object upload size via a policy? Right now, it is set to 50 MB by default in the hosted platform and is not configurable. I have created an issue for this - https://github.com/supabase/storage-api/issues/16

Or are you trying to restrict reading / updating of objects via a policy? This should be possible via policies.

**Reply by Vinzius** [reply]
Hey!

I'm trying to restrict object upload size via a policy, yes :)

I thought that maybe it would be doable with a "insert with check" on the metadata.size. But maybe the object is uploaded before the object can be inserted. Maybe it doesn't make sense to be achieved this way. You'll know better than me :)

Anyway, 50MB is already good for a start!

I'll subscribe to the issue ^^

Thanks!

**Reply by inian** [reply]
Yes, for insert [the rough logic](https://github.com/supabase/storage-api/blob/master/src/routes/object/createObject.ts) is
- insert row into the database
- if successful, start streaming the object to s3. We don't know the size of the object at this point since the object is not buffered in memory (for perf and to keep a small memory footprint)
- after the object is uploaded, we get the object size from s3 and update the metadata in the db

Since the insert happens in step number 1 itself, the policy checking for size metadata won't affect it.
